sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -85,26 +87,103 @@ static RT::PiProgram createSpirvProgram(const RT::PiContext Context,
   return Program;
 }
 
+template <typename ExceptionT, typename RetT>
+RetT *
+waitUntilBuilt(KernelProgramCache &Cache,
+               KernelProgramCache::EntityWithState<RetT> *WithBuildState) {
+  Cache.waitUntilBuilt([WithBuildState]() {
+    auto State = WithBuildState->State.load();
+
+    return State == BS_Done || State == BS_Failed;
+  });
+
+  auto *Result = WithBuildState->Ptr.load();
+
+  if (!Result)
+    throw ExceptionT("The other thread tried to build the program/kernel but "
+                     "did not succeed.");
+
+  return Result;
+}
+
+template <typename RetT, typename ExceptionT, typename KeyT, typename AcquireFT,
+          typename GetCacheFT, typename BuildFT>
+RetT *getBuiltEntity(KernelProgramCache &Cache, const KeyT &CacheKey,
+                     AcquireFT &&Acquire, GetCacheFT &&GetCache,
+                     BuildFT &&Build) {
+  bool InsertionTookPlace;
+  KernelProgramCache::EntityWithState<RetT> *WithState;
+
+  {
+    auto Locked = Acquire(Cache);
```

Comment:
Please, do not use auto when the type is not clear for the line itself. Please, apply to whole patch.

Reply:
Done.
BTW, it is template function. It is intended to not know the exact type here.